gstat/libcsf/man3/Mopen.3
.lf 1 Mopen.3
.\" WARNING! THIS FILE WAS GENERATED AUTOMATICALLY BY c2man!
.\" DO NOT EDIT! CHANGES MADE TO THIS FILE WILL BE LOST!
.TH "Mopen" 3 "13 August 1999" "c2man mopen.c"
.SH "NAME"
Mopen \- open an existing CSF file
.SH "SYNOPSIS"
.ft B
#include "csf.h"
.br
.sp
MAP *Mopen
.br
(
.br
	const char *fileName,
.br
	enum MOPEN_PERM mode
.br
);
.ft R
.SH "PARAMETERS"
.TP
.B "const char *fileName"
file name
.TP
.B "enum MOPEN_PERM mode"
file permission
.sp
Possible values for an \fBenum MOPEN_PERM\fR are as follows:
.RS 0.75in
.PD 0
.ft B
.nr TL \w'M_READ_WRITE'u+0.2i
.ft R
.TP \n(TLu
\fBM_READ\fR
open read only
.TP \n(TLu
\fBM_WRITE\fR
open write only
.TP \n(TLu
\fBM_READ_WRITE\fR
open for both reading and writing
.RE
.PD
.SH "DESCRIPTION"
Mopen opens a CSF file. It allocates space for
the MAP runtime-structure, reads the header file
and performs test to determine if it is a CSF file.
The MinMaxStatus is set to MM_KEEPTRACK if the min/max
header fields are not MV or MM_WRONGVALUE if one of them
contains a MV.
.SH "RETURNS"
a pointer the MAP runtime structure if the file is
successfully opened as a CSF file, NULL if not.
.SH "MERRNO"
NOCORE BADACCESMODE OPENFAILED NOT_CSF BAD_VERSION
.SH "EXAMPLE"
.lf 1 examples/testcsf.tr
.DS
 #include <stdio.h>
 #include "csf.h"
 
 /* test if files are CSF files
  * writing error messages to
  * stdout  
  */
 
 void main(int argc, char *argv[] )
 {
   MAP *map;
   int i;
 
   for(i=1; i < argc; i++)
     if ((map = Mopen(argv[1], M_READ)) == NULL)
     {
         printf("%s %s\\n", argv[i], MstrError());
         ResetMerrno();
     }
     else
     {
         printf("%s is a CSF file\\n",argv[i]);
         Mclose(map);
     }
 
   exit(0);
 }
 
.DE
.lf 60 Mopen.3


src/CHANGES
/*
CHANGES log (summary)
============================
Changes written after a version number apply to the next the version.
date: version (program name)

dec 92, spil 2.06:
	added method ossfim, dump_cvv(); dynamic data maximum.
jan 93, spil 2.07:
	dynamic allocation data, unbounded maximum.
mar 93, spil 2.08:
	corrected block discretizing
apr 93, spil 2.09:
	added lex variogram expression parsing (rd_vario.[lc]) ## later removed
apr 93, spil 3.0:
	added full command file parser with lexical analyser
	added sample variogram calculation (sem)

jun 93, gstat 0.99
	spil 3.0 + sem -> gstat 0.99; "gstat [-]sem" ## later: -e semivariance

jun 93, gstat 0.99c:
	removed bug in local/ordinary kriging
jun 93, gstat 0.99d:
	coupled gstat with meschach library; removed numerical recip matrix code.
jun 93, gstat 0.99e:
	solved bug in semi().
aug 93, gstat 0.99f:
	added MapSign().
aug 93, gstat 0.99g:
	added time coordinate to data, distance(x,t) not solved.
	corrected distance bug when d->force in select.c
	added variogram->max_[range/sill] 
aug 93, gstat 0.99h
	added max_ldl structure, and correction with -L option
sep 93, gstat 0.99i
	dynamic allocation of nested variogram models (unbounded maximum)
sep 93, gstat 0.99j
	added universal kriging
oct 93, gstat 1.0  
	release without univ.kr., added variography check (ic,lmc).
nov 93, gstat 1.1  
	renewed error handling, corrected variography check
	corrected bug in memory allocation in ReadLegend();
	released gstat 1.1 on dec 1, 1993

dec 93, gstat 1.2 
	changed the VARIOGRAM structure, using V_model and V_model_type;
	[ check UK, consider REML? ]
	added 4x4 Gauss quadrature block descretizing (Carr & Palmer)
	removed the attr2 field from POINT, changed semi() accordingly
	removed bug in positioning results from ID, if n_vars > 1
	changed UK, added LM. def: with intercept, other variables with X
	revised sem.c , added sem_main.c, made CheckX();
apr 94, gstat 1.2a 
	removed bug in sem_main.c (log, and I for data2);
	removed bug in do_mask.c, warning for coordinates when PT_XY
	removed bug in direct.c (117), added ER_DOMAIN
apr 94, gstat 1.2b
	added gram_sch.c: Gram_Schmidt checking of matrices: option -g
	added # comment-style and flex -l compatability (GNU flex 2.4.6)
	removed bug in header cross semivariogram (pseudo <--> symmetric)
	changed Quant() and Quantile, so they equal Splus' quantile()
june 94, gstat 1.2c
	removed bug when using both Valid and XValid (reversed order)
	removed method tin, lss, ... added method xy
june 94, gstat 1.2d
	removed bug in Skrige() (Oops, I forgot to subtract sk_mean)
june 94, gstat 1.2e
	added reading `points(xxx):"map.csf"'; points from csf map.
    released 1.2f
june 94, gstat 1.2g 
	corrected bug in lm.c: pred. variance, depending on block size
	placed '#' on start of the 8 header lines of semivariograms, 
	to make life easier for gnuplot support
aug 94, gstat 1.2h
	added 'par:fit=1,gnuplot="file.gp"; functionality: see sem.c, 
	sem_fit.c this is for wls fitting of sill-nugget, and variogram 
	plotting using gnuplot. (added # in variogram header!)
aug 94, gstat 1.3 
	Added mivque.[ch], for mi[nv]que univariate covariance fitting
aug 94, gstat 1.3a
	added global kriging in [s]krige.c. Up to 80 times faster. Wow.
aug 94, gstat 1.3b
	added `area: "discr.dat",x=1,y=2', user defined area discretising
sep 94, gstat 1.3c
	fixed bug in CovSem_AB(): remember both within-block cov. and sem.
oct 94, gstat 1.3d
	added menu; corrected CheckX() for points();, but see 1.4b...
	added noaverage/average instead of "average=1"
	added "square" to use square neigborhood selections (no circle)
	changed variogram structure, now functions as part of structure.
	removed bug in getest, functions for MVS/MED, added par:quantile=..
dec 94, gstat 1.3e 
	menu evolved... added '#','?','-','&' options (try ?);
	added recalculation only at options Data, Calculate, cUtoff, &

----------gstat, 1995:
jan 94, gstat 1.4  
	menu.c elm-like, major rewrite. No hidden options, help added.  
jan 94, gstat 1.4a 
	corrected compile with STRICT_KR2 (no HAVE_CURSES)
	menu faster (incomplete screen rewrite after each key pressed)
	env. variable GNUPROLOG may define a prolog for gnuplot
jan 94, gstat 1.4b 
	corrected bug in select.c (cs with no range defined)
	env. variable GNUPLOT points gnuplot out, GSTATFMT "%g" variogram
	menu enhancements (echo Yes/No, Show)
	Yet another nasty bug in CheckX()!
feb 94, gstat  1.4c 
	header of variogram cloud corrected. Found bug in mopen.c (csf-lib)
	IsEas(): now use rewind() instead of fopen()+fclose().
	partial fitting: 0.2 @Nug() + 1 Sph(30) will now only fit Sph part.
	(removed obsolete Wls_fit_Sill())
	-L : allow local residuals.
	added dX to DATA: max X-distance allowed for semivar. calculation
	(removed obsolete `factor' from DATA. This can now be done
	in a more general fashion with X and dX).
	menu options: 'J': jgraph file; `&' jgraph-eps; '!': shell command.
Mar 94, gstat 1.4d 
	cleaned up random.c, removed rn.[ch] and Ran[0-3]()
	corrected estimation of sample variance (lm->s2) for weighted lm.
 	EnterData() (menu.c) now does not return without valid data read.
	temporary file names now come from tmpnam() or tempnam() [on UNIX];
	added MODE mode: more flexibility now (stratify whenever suitable)
	Fixed bug in Xvalid(): where.X, NA's on uk/lm for secondary vars
	Fixed bug (weights) in CalcLM(); added 's', and '+/-' in menu.
	Added DATE in date.h, changed program heading: -V now includes date.
	Now mean variances (V=..) are subtracted in variogram calculation.
Apr 94, gstat 1.4.e
	Made lm->SSR the corrected SS for mean, with p-1 df (Christensen p. 89)
	Added more.c (callable from menu). (removed later on...)
	syntax change: `pars:x=value;' to `set x=value;' (like gnuplot)
	Made `width' and `cutoff' global variables (`set cutoff=15,width=1').
	fit.c: now use lm.c routines for the fitting (estimation),
	Changed convergence criterion in WlsFit(), Mivque():

libcsf/src/mopen.c

#ifndef lint
 static const char *rcs_id = 
 "$Header: /home/pcrtree/SRC.RCS/libs/csf/RCS/mopen.c,v 2.1 1996/12/29 19:35:21 cees Exp $";
#endif

/*******************************************************/
/*	FUNCTIE   MOPEN.C	                       */
/*******************************************************/
/*******************************************************/


#include <string.h>

#include "csf.h"
#include "csfimpl.h"


static const char *openModes[3] = {
	S_READ,
	S_WRITE,
	S_READ_WRITE
	};

/* open an existing CSF file
 * Mopen opens a CSF file. It allocates space for
 * the MAP runtime-structure, reads the header file
 * and performs test to determine if it is a CSF file.
 * The MinMaxStatus is set to MM_KEEPTRACK if the min/max
 * header fields are not MV or MM_WRONGVALUE if one of them
 * contains a MV.
 * returns a pointer the MAP runtime structure if the file is
 * successfully opened as a CSF file, NULL if not.
 *
 * Merrno
 * NOCORE BADACCESMODE OPENFAILED NOT_CSF BAD_VERSION
 *
 * EXAMPLE
 * .so examples/testcsf.tr
 */
MAP  *Mopen(
	const char *fileName, /* file name */
	enum MOPEN_PERM mode) /*  file permission */
{
 MAP *m;
 UINT4 s; /* swap detection field */
 
 if (! CsfIsBootedCsfKernel())
 	CsfBootCsfKernel();
 
 m = (MAP *)CSF_MALLOC(sizeof(MAP));
 
 if (m == NULL)
 {
 	M_ERROR(NOCORE);
 	goto error_mapMalloc;
 }
 
 m->fileName = (char *)CSF_MALLOC(strlen(fileName)+1);
 if (m->fileName == NULL)
 {
 	M_ERROR(NOCORE);
 	goto error_fnameMalloc;
 }
 (void)strcpy(m->fileName,fileName);
 
 /* check file mode validation */
 if ( IS_BAD_ACCESS_MODE(mode))
 {
 	M_ERROR(BADACCESMODE);
 	goto error_notOpen;
 }
 m->fileAccessMode = mode;
 
 
 /*  check if file can be opened or exists */
 m->fp = fopen(fileName, openModes[mode-1]);
 if (m->fp == NULL)
 {
 	M_ERROR(OPENFAILED);
 	goto error_notOpen;
 }
 
 /*  check if file could be C.S.F.-file 
  *   (at least 256 bytes long) 
  *  otherwise the signature comparison will
  *  fail
  */
 
 (void)fseek(m->fp,0L, SEEK_END);
 if (ftell(m->fp) < (long)ADDR_DATA)
 {
 	M_ERROR(NOT_CSF);
 	goto error_open;
 }

 (void)fseek(m->fp, 14+CSF_SIG_SPACE, SEEK_SET);
 (void)fread((void *)&s, sizeof(UINT4),1,m->fp);
 if (s != ORD_OK) {
	m->write = CsfWriteSwapped;
	m->read  = CsfReadSwapped;
 }
 else {
#ifdef DEBUG
	m->read  = (CSF_READ_FUNC)CsfReadPlain;
	m->write = (CSF_READ_FUNC)CsfWritePlain;
#else
	m->read  = (CSF_READ_FUNC)fread;
	m->write = (CSF_READ_FUNC)fwrite;
#endif
 }
 
 (void)fseek(m->fp, ADDR_MAIN_HEADER, SEEK_SET);
 m->read((void *)&(m->main.signature), sizeof(char), CSF_SIG_SPACE,m->fp);
 m->read((void *)&(m->main.version),   sizeof(UINT2),1,m->fp);
 m->read((void *)&(m->main.gisFileId), sizeof(UINT4),1,m->fp);
 m->read((void *)&(m->main.projection),sizeof(UINT2),1,m->fp);
 m->read((void *)&(m->main.attrTable), sizeof(UINT4),1,m->fp);
 m->read((void *)&(m->main.mapType),  sizeof(UINT2),1,m->fp);
 m->read((void *)&(m->main.byteOrder), sizeof(UINT4),1,m->fp);
 /*                                             14+CSF_SIG_SPACE
  */
 
 (void)fseek(m->fp, ADDR_SECOND_HEADER, SEEK_SET);
 m->read((void *)&(m->raster.valueScale), sizeof(UINT2),1,m->fp);
 m->read((void *)&(m->raster.cellRepr), sizeof(UINT2),1,m->fp);

 (void)fread((void *)&(m->raster.minVal), sizeof(CSF_VAR_TYPE), 1,m->fp);
 (void)fread((void *)&(m->raster.maxVal), sizeof(CSF_VAR_TYPE), 1,m->fp);
 if (s != ORD_OK) {
  CsfSwap((void *)&(m->raster.minVal), CELLSIZE(m->raster.cellRepr), 1);
  CsfSwap((void *)&(m->raster.maxVal), CELLSIZE(m->raster.cellRepr), 1);
 }

 m->read((void *)&(m->raster.xUL), sizeof(REAL8),1,m->fp);
 m->read((void *)&(m->raster.yUL), sizeof(REAL8),1,m->fp);
 m->read((void *)&(m->raster.nrRows), sizeof(UINT4),1,m->fp);
 m->read((void *)&(m->raster.nrCols), sizeof(UINT4),1,m->fp);
 m->read((void *)&(m->raster.cellSizeX), sizeof(REAL8),1,m->fp);
 m->read((void *)&(m->raster.cellSizeY), sizeof(REAL8),1,m->fp);
 
 m->read((void *)&(m->raster.angle), sizeof(REAL8),1,m->fp);
 
 
 /*  check signature C.S.F.file	
  */
 if(strncmp(m->main.signature,SIG,SIZE_SIG)!=0)
 {
 	M_ERROR(NOT_CSF);
 	goto error_open;
 }
 /* should be read right
  */
 POSTCOND(m->main.byteOrder == ORD_OK);
 /*  restore byteOrder C.S.F.file (Intel or Motorola)  */
 m->main.byteOrder=s;
 
 /*  check version C.S.F.file	
  */
 if (m->main.version != CSF_VERSION_1
    && (m->main.version != CSF_VERSION_2))
 {
 	M_ERROR(BAD_VERSION);
 	goto error_open;
 }
 
 if (m->main.version == CSF_VERSION_1)
 	m->raster.angle = 0.0;
 
 CsfSetMapAngleCosSin(m);
 
 CsfRegisterMap(m);
 
 /* install cell value converters: (app2file,file2app)
  */
 m->app2file = CsfDummyConversion;
 m->file2app = CsfDummyConversion;
 m->appCR    = m->raster.cellRepr;
 
 if (IsMV(m,&(m->raster.minVal)) ||
     IsMV(m,&(m->raster.maxVal))   )
 	m->minMaxStatus = MM_WRONGVALUE;
 else
 	m->minMaxStatus = MM_KEEPTRACK;
 
 return(m);
 
 error_open:
 PRECOND(m->fp != NULL);
 (void)fclose(m->fp);
 error_notOpen: 
 CSF_FREE(m->fileName);
 error_fnameMalloc: 
 CSF_FREE(m);
error_mapMalloc:
	 return(NULL);
}
